source/abstract/memory/ExpHeap.cc

```diff
@@ -0,0 +1,346 @@
+        prev->m_link.m_next = block;
+    } else {
+        next = m_head;
+        m_head = block;
+    }
+
+    block->m_link.m_next = next;
+    if (next) {
+        next->m_link.m_prev = block;
+    } else {
+        m_tail = block;
+    }
+
+    return block;
+}
+
+MEMiExpBlockHead *MEMiExpBlockList::append(MEMiExpBlockHead *block) {
+    return insert(block, m_tail);
+}
+
+MEMiExpBlockHead *MEMiExpBlockList::remove(MEMiExpBlockHead *block) {
+    MEMiExpBlockHead *prev = block->m_link.m_prev;
+    MEMiExpBlockHead *next = block->m_link.m_next;
+
+    if (prev) {
+        prev->m_link.m_next = next;
+    } else {
+        m_head = next;
+    }
+
+    if (next) {
+        next->m_link.m_prev = prev;
+    } else {
+        m_tail = prev;
+    }
+
+    return prev;
+}
+
+MEMiExpBlockHead::MEMiExpBlockHead(const Region &region, u16 signature) {
+    m_signature = signature;
+    m_attribute.val = 0;
+
+    m_size = region.getRange() - sizeof(MEMiExpBlockHead);
+
+    m_link.m_prev = nullptr;
+    m_link.m_next = nullptr;
+}
+
+MEMiExpBlockHead *MEMiExpBlockHead::createFree(const Region &region) {
+    constexpr u16 FREE_BLOCK_SIGNATURE = 0x4652; // FR
+    return new (region.start) MEMiExpBlockHead(region, FREE_BLOCK_SIGNATURE);
+}
+
+MEMiExpBlockHead *MEMiExpBlockHead::createUsed(const Region &region) {
+    constexpr u16 USED_BLOCK_SIGNATURE = 0x5544; // UD
+    return new (region.start) MEMiExpBlockHead(region, USED_BLOCK_SIGNATURE);
+}
+
+Region MEMiExpBlockHead::getRegion() const {
+    return Region(SubOffset(this, m_attribute.fields.alignment), getMemoryEnd());
+}
+
+void *MEMiExpBlockHead::getMemoryStart() const {
+    return AddOffset(this, sizeof(MEMiExpBlockHead));
+}
+
+void *MEMiExpBlockHead::getMemoryEnd() const {
+    return AddOffset(getMemoryStart(), m_size);
+}
+
+MEMiExpHeapHead::MEMiExpHeapHead(void *end, u16 opt)
+    : MEMiHeapHead(EXP_HEAP_SIGNATURE, AddOffset(this, sizeof(MEMiExpHeapHead)), end, opt) {
+    m_groupId = 0;
+    m_attribute.makeAllZero();
+
+    Region region = Region(getHeapStart(), getHeapEnd());
+    MEMiExpBlockHead *block = MEMiExpBlockHead::createFree(region);
+
+    m_freeBlocks.m_head = block;
+    m_freeBlocks.m_tail = block;
+    m_usedBlocks.m_head = nullptr;
+    m_usedBlocks.m_tail = nullptr;
+}
+
+MEMiExpHeapHead::~MEMiExpHeapHead() = default;
+
+MEMiExpHeapHead *MEMiExpHeapHead::create(void *startAddress, size_t size, u16 flag) {
+    void *endAddress = AddOffset(startAddress, size);
+
+    startAddress = RoundUp(startAddress, 4);
+    endAddress = RoundDown(endAddress, 4);
+
+    uintptr_t startAddrNum = GetAddrNum(startAddress);
+    uintptr_t endAddrNum = GetAddrNum(endAddress);
+
+    if (startAddrNum > endAddrNum) {
+        return nullptr;
+    }
+
+    if (endAddrNum - startAddrNum < sizeof(MEMiExpHeapHead) + sizeof(MEMiExpBlockHead) + 4) {
+        return nullptr;
+    }
+
+    return new (startAddress) MEMiExpHeapHead(endAddress, flag);
+}
+
+void MEMiExpHeapHead::destroy() {
```

Comment:
```suggestion
/// @addr{0x80198D58}
void MEMiExpHeapHead::destroy() {
```

Reply:
This function doesn't exist in the base game. It's a helper to call the destructor, where the destructor would be analogous to `MEMDestroyExpHeap`.